✨ Parse command line arguments to pass into the function

ThePhD opened the pull request:
We still need to handle things like stdin and stdout and input/output, but that can be done a little later. None of the variables are used and reacted on properly but it will help us make progress. Likely conflicts with #3, so this one should likely be merged first!

ThePhD committed:
✨ Parse command line arguments to pass into the function

— 💬 "We still need to handle things like stdin and stdout and input/output, but that can be done a little later!"

mcy commented on Cargo.toml:
```diff
@@ -1,22 +1,26 @@
+
+[dependencies]
+[dependencies.bumpalo]
+version = ">=0,<99999999999999"
```
`= "*"` is probably what you want. :P

ThePhD replied:
="*" *is* what I want, but if I do that then apparently cargo will be all >:( angry-face about it if we ever upload this. (We probably won't but I'm doing this on pricinple; I can probably change it back. :D)

mcy commented on badc/source/main.rs:
```diff
@@ -1,21 +1,44 @@
+#[command(author, version, about = "A B language (Kernighan, 1969) compiler. Not at all useful.")]
+struct CommandLineCompilationOptions {
+	/// All of the paths to the input to compile, each one considered an independent translation unit.
+	#[arg(short, long)]
```
Hmm, I think that for `badc foo.b` to work, you want to not set `short` or `long`, right?

ThePhD replied:
Right; I think that's how it does work. Removing!

mcy commented on badc/source/main.rs:
```diff
@@ -1,21 +1,44 @@
+	}
+	for input in args.inputs {
+		let output = match args.output {
+			Some(ref target_path) => target_path.clone(),
```
`ref` should be redundant here?

ThePhD replied:
The error says:

> use of moved value
> move occurs because value has type `PathBuf`, which does not implement the `Copy` traitrustc[E0382](https://doc.rust-lang.org/error-index.html#E0382)
> main.rs(32, 9): borrow this field in the pattern to avoid moving `args.output.0`: `ref `

So I think I have to add the `ref`...

ThePhD replied:
Fixed it!!!

mcy commented on badc/source/main.rs:
```diff
@@ -1,21 +1,44 @@
-	println!("{:?} 🎉!", tree);
+	let mut args = CommandLineCompilationOptions::parse();
+	if args.inputs.is_empty() {
+		args.inputs.push(PathBuf::from_str("./main.b").unwrap());
```
You can just use `from` here, which is infallible.

ThePhD replied:
Done ✨

ThePhD committed:
✨ Token printing output paths

— 💬 "As a TODO, we should probably make it so we serialize more directly into ProgramSink, so we can pass stdout/stdin as options through a single dash (-)..."

ThePhD committed:
🎨 Format available files to get proper, approved structure.

ThePhD commented:
@mcy Good to go now that I've cleaned that stuff up? I promise I'll figure out WTF is going on with the "ref" situation, but that'll come afterwards if that's okay.

(EDIT: I don't mean to over-pester or anything, just to make it clear I don't have anything else to add to this jwakdwjad. Maybe I should have said it differently...)

ThePhD committed:
🛠 Use & on the input instead of "ref" on the match.

— 💬 "Idiomatic Rust, bam!! 💥 "

ThePhD commented:
Fixed everything!!!

ThePhD committed:
🛠 More use of & on the input instead of "ref" on the match.

— 💬 "More Idiomatic Rust, bam!! 💥 "
